relay-cardinality/src/redis/limiter.rs

```diff
@@ -147,164 +145,32 @@ impl Limiter for RedisSetLimiter {
         let mut connection = client.connection()?;
 
         for mut state in states {
-            if state.entries.is_empty() {
-                continue;
-            }
-
-            let results = metric!(timer(CardinalityLimiterTimers::Redis), id = state.id(), {
-                self.check_limits(&mut connection, &mut state, timestamp)
-            })?;
-
-            // This always acquires a write lock, but we only hit this
-            // if we previously didn't satisfy the request from the cache,
-            // -> there is a very high chance we actually need the lock.
-            let mut cache = self.cache.update(state.scope, timestamp); // Acquire a write lock.
-            for (entry, status) in results {
-                if status.is_rejected() {
-                    rejections.reject(state.cardinality_limit(), entry.id);
-                    state.rejected();
-                } else {
-                    cache.accept(entry.hash);
-                    state.accepted();
+            for (scope, entries) in state.take_scopes() {
+                let results = metric!(timer(CardinalityLimiterTimers::Redis), id = state.id(), {
+                    self.check_limits(&mut connection, &state, scope, entries, timestamp)
+                })?;
+
+                // This always acquires a write lock, but we only hit this
+                // if we previously didn't satisfy the request from the cache,
+                // -> there is a very high chance we actually need the lock.
+                let mut cache = self.cache.update(scope, timestamp); // Acquire a write lock.
+                for (entry, status) in results {
+                    if status.is_rejected() {
+                        rejections.reject(state.cardinality_limit(), entry.id);
+                        state.rejected();
+                    } else {
+                        cache.accept(entry.hash);
+                        state.accepted();
+                    }
                 }
+                drop(cache); // Give up the cache lock!
             }
-            drop(cache); // Give up the cache lock!
         }
 
         Ok(())
     }
 }
 
-/// Internal state combining relevant entries for the respective quota.
-struct LimitState<'a> {
```

Comment:
(I find moving this struct with the rest of the changes in the PR makes tracking actual changes quite hard)

Reply:
Agree with this. Better to split PRs into "this only moves stuff around" and a separate one for functional changes.

Reply:
Done in #3329